### PR TITLE
feat(serve): deploy hardware information collector for worker nodes

### PR DESCRIPTION
**Pull request by JustinTong0323**
## Description
This PR introduces a **Hardware Information Collector** for worker nodes within the cluster. The expected behavior is as follows:
- Automatic Execution: When initializing the cluster, the Hardware Information Collector will automatically run on each new worker node.

### Roadmap:
- [x] Implement the initial script responsible for gathering hardware metrics from worker nodes.
- [x] Utilize Ray to execute the hardware collection script on worker nodes remotely.
- [x] (Not automated) Ensure the hardware information collector script runs specifically on each newly added worker node.
- [x] (Not automated) Save the collected hardware information to the `hardware_config.json` file on the control node for future reference and scheduling decisions.
- [x] Integrated into the controller for automated operation.

## Motivation
Collecting detailed hardware profiles of worker nodes is crucial for implementing **Storage Aware Scheduling**. Accurate and up-to-date hardware information enables the scheduler to make informed decisions, optimizing resource allocation based on each node’s capabilities and current state.

## Type of Change
- [x] New feature

## Checklist
- [x] I have read the [CONTRIBUTION](https://github.com/ServerlessLLM/ServerlessLLM/blob/main/CONTRIBUTING.md) guide.
- [ ] I have updated the tests (if applicable).
- [x] I have updated the documentation (if applicable).

### CODE REVIEW

**Review by future-xy — COMMENTED** [review]
Could you make sure that after applying this pr the storage aware scheduling will still work? and update the [docs](https://serverlessllm.github.io/docs/stable/serve/storage_aware_scheduling)

File: docs/stable/serve/storage_aware_scheduling.md
Comment (future-xy) [inline docs/stable/serve/storage_aware_scheduling.md:1]: Is it possible to update this document to use docker compose?

**Review by future-xy — CHANGES_REQUESTED** [review]
Overall this pr looks good to me. However, a few comments before merge:
1. Could you merge the latest main (when #148 is merged) to fix the Transformers loading bug?
2. Could you update the documents to use docker compose for test?
3. Please see other comments left in the reviews.

File: sllm/serve/hardware_info_collector.py [hunk 1]
```diff
@@ -0,0 +1,331 @@
...
+import speedtest
+from tqdm import tqdm
+
+from sllm.serve.logger import init_logger
+
+logger = init_logger(__name__)
+
+
+@ray.remote
+class HardwareInfoCollector:
```
[the remark this answers is not in the document]
Comment (JustinTong0323) [reply sllm/serve/hardware_info_collector.py:19]: Yes. As we need to using several isolated methods to collect different hardware configs, using actor is necessary to pass these methods remotely to the worker nodes.
It's also beneficial for further extensibility, like checks for the adding and removing of worker nodes.

File: sllm/serve/hardware_info_collector.py [hunk 2]
```diff
@@ -0,0 +1,331 @@
...
+    Args:
+        benchmark_results (Dict[str, dict]): The benchmark results mapped by node IP.
+        filename (str): The filename for the JSON output.
+    """
+    with open(filename, "w") as f:
+        json.dump(benchmark_results, f, indent=4)
+    logger.info(f"Benchmark results stored in {filename}")
+
+
+def main():
```
[the remark this answers is not in the document]
Comment (JustinTong0323) [reply sllm/serve/hardware_info_collector.py:260]: Yes I can remove it.

File: sllm/serve/hardware_info_collector.py [hunk 3]
```diff
@@ -0,0 +1,331 @@
...
+        hardware_info["disk_read_bandwidth"] = read_bw
+        # Temporarily disabled due to network issues
+        # upload_bw, download_bw = self.collector.get_network_bandwidth()
+        # hardware_info["network_upload_bandwidth"] = upload_bw
+        # hardware_info["network_download_bandwidth"] = download_bw
+        hardware_info["GPUs_info"] = self.collector.get_gpu_info()
+        return hardware_info
+
+
+class HardwareInfoCollectorMethods:
```
[the remark this answers is not in the document]
Comment (JustinTong0323) [reply sllm/serve/hardware_info_collector.py]: Yes that's correct, using a class is not necessary.  Already refactored.

File: sllm/serve/hardware_info_collector.py [hunk 4]
```diff
@@ -0,0 +1,331 @@
...
+    A Ray actor class responsible for collecting hardware information.
+    """
+
+    def __init__(self):
+        """
+        Initialize the HardwareInfoCollector actor.
+        """
+        self.collector = HardwareInfoCollectorMethods()
+
+    def collect_all_info(self):
```
[the remark this answers is not in the document]
Comment (JustinTong0323) [reply sllm/serve/hardware_info_collector.py:24]: Yes, already refactored.

File: sllm/serve/hardware_info_collector.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JustinTong0323) [reply sllm/serve/hardware_info_collector.py:19]: Just Refactored.

File: docs/stable/serve/storage_aware_scheduling.md
Comment (JustinTong0323) [reply docs/stable/serve/storage_aware_scheduling.md:1]: Done.

**Review by future-xy — CHANGES_REQUESTED** [review]
I've tested this PR, and it works well for me. Could you please move the Docker Compose file, along with the two configuration JSON files, into an `examples` directory? This adjustment is necessary because the Docker Compose file is context-sensitive (see [example here](https://github.com/ServerlessLLM/ServerlessLLM/blob/de88c8a06a98c5b10315f6ea189cc44c8e0a638c/examples/docker/docker-compose.yml#L21)).

Thanks!

**Comment by JustinTong0323** [comment]
> I've tested this PR, and it works well for me. Could you please move the Docker Compose file, along with the two configuration JSON files, into an `examples` directory? This adjustment is necessary because the Docker Compose file is context-sensitive (see [example here](https://github.com/ServerlessLLM/ServerlessLLM/blob/de88c8a06a98c5b10315f6ea189cc44c8e0a638c/examples/docker/docker-compose.yml#L21)).
> 
> Thanks!

Do you mean create a new directory like `examples/storage_aware_scheduling` and put model config files and docker compose config in it? In that case the doc could just reference to the config, is that right?

**Comment by future-xy** [comment]
> > I've tested this PR, and it works well for me. Could you please move the Docker Compose file, along with the two configuration JSON files, into an `examples` directory? This adjustment is necessary because the Docker Compose file is context-sensitive (see [example here](https://github.com/ServerlessLLM/ServerlessLLM/blob/de88c8a06a98c5b10315f6ea189cc44c8e0a638c/examples/docker/docker-compose.yml#L21)).
> > Thanks!
> 
> Do you mean create a new directory like `examples/storage_aware_scheduling` and put model config files and docker compose config in it? In that case the doc could just reference to the config, is that right?

Yes, exactly! Also, please update the documentation to instruct users to clone the repository and `cd` into that directory for setup.

File: docs/stable/serve/storage_aware_scheduling.md [hunk 5]
```diff
@@ -2,135 +2,67 @@
...
 ```bash
-docker run -d --name ray_head \
-  --runtime nvidia \
-  --network sllm \
-  -p 6379:6379 \
-  -p 8343:8343 \
-  --gpus '"device=none"' \
-  serverlessllm/sllm-serve
+git clone https://github.com/ServerlessLLM/ServerlessLLM.git
+cd serverlessllm/examples/storage_aware_scheduling
```
Comment (future-xy) [inline docs/stable/serve/storage_aware_scheduling.md:21]: There is an inconsistency in the path naming: please standardize to "ServerlessLLM" for consistency across the project.

**Review by future-xy — CHANGES_REQUESTED** [review]

File: docs/stable/serve/storage_aware_scheduling.md (hunk 5 above)
Comment (JustinTong0323) [reply docs/stable/serve/storage_aware_scheduling.md]: Modified. I just follow the pattern in Docker Quickstart Guide, so I guess another pr is needed to modify that.

Comment (JustinTong0323) [reply docs/stable/serve/storage_aware_scheduling.md]: Modified. I just follow the pattern in Docker Quickstart Guide, so I open another pr #154 to modify that.

**Review by future-xy — APPROVED** [review]

**Pull request merged**